### PR TITLE
feat(semantic): add ClassTable

### PR DESCRIPTION
**Pull request by Dunqing**

### CODE REVIEW

**Comment by Dunqing** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/1793?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `main`
  * **PR #1792** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/1792?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #1793** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/1793?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/oxc-project/oxc/1793?utm_source=stack-comment).

File: crates/oxc_semantic/src/builder.rs [hunk 1]
```diff
@@ -62,6 +67,11 @@ pub struct SemanticBuilder<'a> {
     pub current_node_flags: NodeFlags,
     pub current_symbol_flags: SymbolFlags,
     pub current_scope_id: ScopeId,
+    // class
+    pub current_class_id: Option<ClassId>,
+    pub current_property_id: PropertyId,
+    pub current_method_id: MethodId,
```
Comment (Boshen) [inline crates/oxc_semantic/src/builder.rs:73]: Let's move all implementation details into a `ClassTableBuilder`.

File: crates/oxc_semantic/tests/util/class_terser.rs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        });
+        ClassTester {
+            semantic: Rc::new(semantic),
+            class_id: class_id.unwrap_or_else(|| panic!("Cannot find {name} class")),
+        }
+    }
+
+    pub fn has_number_of_properties(&self, len: usize) -> &Self {
+        let property_len = self.semantic.classes().properties[self.class_id].len();
+        assert!(property_len == len, "Expected {len} properties, found {property_len}");
```
Comment (Boshen) [inline crates/oxc_semantic/tests/util/class_terser.rs:34]: let's use `debug_assert` for all these.

File: crates/oxc_semantic/tests/util/class_terser.rs
Comment (Boshen) [inline crates/oxc_semantic/tests/util/class_terser.rs:1]: terser -> tester

File: crates/oxc_semantic/tests/util/mod.rs [hunk 3]
```diff
@@ -108,12 +110,22 @@ impl SemanticTester {
         SymbolTester::new_at_root(self, self.build(), name)
     }
 
+    /// Tests that a class with the given name exists
+    ///
+    /// ## Fails
+    /// If no class with the given name exists.
+    #[allow(dead_code)]
```
Comment (Boshen) [inline crates/oxc_semantic/tests/util/mod.rs:117]: Do we need dead_code in the tester?

**Review by Boshen — CHANGES_REQUESTED** [review]

**Comment by Boshen** [comment]
There is no performance change because our benchmark suite are plain es5 files, we need to improve them (change some of the files to larger typescript files that uses more ecmascript features) to better reflect performance changes.

File: crates/oxc_semantic/tests/util/mod.rs (hunk 3 above)
Comment (Dunqing) [reply crates/oxc_semantic/tests/util/mod.rs:117]: This line can be deleted, but dead_code is needed in ClassTester. I don't know why
image

Comment (Boshen) [reply crates/oxc_semantic/tests/util/mod.rs:117]: Oh, they are only used in #[test] ... try add #[test] to these methods?

**Comment by Boshen** [comment]
Just need to pass the test now?

File: crates/oxc_semantic/src/builder.rs [hunk 4]
```diff
@@ -84,6 +85,7 @@ pub struct SemanticBuilder<'a> {
     check_syntax_error: bool,
 
     redeclare_variables: RedeclareVariables,
+    class: ClassTableBuilder,
```
Comment (Boshen) [inline crates/oxc_semantic/src/builder.rs:88]: `class_table_builder` to make things more readable?

File: crates/oxc_semantic/src/builder.rs [hunk 5]
```diff
@@ -169,6 +172,7 @@ impl<'a> SemanticBuilder<'a> {
             nodes: self.nodes,
             scopes: self.scope,
             symbols: self.symbols,
+            classes: self.class.classes,
```
Comment (Boshen) [inline crates/oxc_semantic/src/builder.rs:175]: better to have a `.build` method.

File: crates/oxc_semantic/src/builder.rs [hunk 6]
```diff
@@ -398,6 +403,8 @@ impl<'a> Visit<'a> for SemanticBuilder<'a> {
 
 impl<'a> SemanticBuilder<'a> {
     fn enter_kind(&mut self, kind: AstKind<'a>) {
+        // let parent_node = self.nodes.parent_node(self.current_node_id);
```
Comment (Boshen) [inline crates/oxc_semantic/src/builder.rs:406]: commented code?

File: crates/oxc_semantic/src/builder.rs [hunk 7]
```diff
@@ -398,6 +403,8 @@ impl<'a> Visit<'a> for SemanticBuilder<'a> {
...
+        self.class.enter_kind(kind, self.current_node_id, &self.nodes);
```
Comment (Boshen) [inline crates/oxc_semantic/src/builder.rs:407]: I think it's better to move and slit these into code blow so we get 1 less AST node check.

Remember, the code is called on everything single ast node - every cpu instruction counts.

**Review by Boshen — APPROVED** [review]

**Pull request merged**